#  DSC530-T304 Data Exploration and Analysis - Week 9 #

## Regression ##

### Exercise Chapter 11  ###

In [1]:
# Download pregency and respondent data files from thinkstat

from os.path import basename, exists

# Declare a function to download the files to local
def download(fileurl):
    filename = basename(fileurl)
    if not exists(filename):
        from urllib.request import urlretrieve

        local, _ = urlretrieve(fileurl, filename)
        print("Downloaded " + local)

In [2]:
# Dowload the python and data files to load pregnency and respondent data 
download("https://github.com/AllenDowney/ThinkStats2/raw/master/code/thinkstats2.py")
download("https://github.com/AllenDowney/ThinkStats2/raw/master/code/thinkplot.py")
download("https://github.com/AllenDowney/ThinkStats2/raw/master/code/nsfg.py")
download("https://github.com/AllenDowney/ThinkStats2/raw/master/code/2002FemPreg.dct")

In [3]:
download(
    "https://github.com/AllenDowney/ThinkStats2/raw/master/code/2002FemPreg.dat.gz"
)
download("https://github.com/AllenDowney/ThinkStats2/raw/master/code/2002FemResp.dct")
download("https://github.com/AllenDowney/ThinkStats2/raw/master/code/2002FemResp.dat.gz")

## Chapter 11 ##
-------

**Exercise 11-1**
Suppose one of your co-workers is expecting a baby and you are participating in an office pool to predict the date of birth. Assuming that bets are placed during the 30th week of pregnancy, what variables could you use to make the best prediction? You should limit yourself to variables that are known before the birth, and likely to be available to the people in the pool.

everpreg / pregnum
height
npostsmk -  smoked during pregnancy
agescrn
ovuprob


In [4]:
import nsfg
#Load the Pregnancy data
preg_df = nsfg.ReadFemPreg()

In [5]:
preg_df.head()

caseid  pregordr  howpreg_n  howpreg_p  moscurrp  nowprgdk  pregend1  \
0       1         1        NaN        NaN       NaN       NaN       6.0   
1       1         2        NaN        NaN       NaN       NaN       6.0   
2       2         1        NaN        NaN       NaN       NaN       5.0   
3       2         2        NaN        NaN       NaN       NaN       6.0   
4       2         3        NaN        NaN       NaN       NaN       6.0   

   pregend2  nbrnaliv  multbrth  ...  laborfor_i  religion_i  metro_i  \
0       NaN       1.0       NaN  ...           0           0        0   
1       NaN       1.0       NaN  ...           0           0        0   
2       NaN       3.0       5.0  ...           0           0        0   
3       NaN       1.0       NaN  ...           0           0        0   
4       NaN       1.0       NaN  ...           0           0        0   

       basewgt  adj_mod_basewgt      finalwgt  secu_p  sest  cmintvw  \
0  3410.389399      3869.349602   6448.271112       2     9      NaN   
1  3410.389399      3869.349602   6448.271112       2     9      NaN   
2  7226.301740      8567.549110  12999.542264       2    12      NaN   
3  7226.301740      8567.549110  12999.542264       2    12      NaN   
4  7226.301740      8567.549110  12999.542264       2    12      NaN   

   totalwgt_lb  
0       8.8125  
1       7.8750  
2       9.1250  
3       7.0000  
4       6.1875  

[5 rows x 244 columns]

In [6]:
# Filter only live birth rows into a data fraem
live_df = preg_df[preg_df.outcome == 1]

From the Excercise question following are identified<br>
**Independent / predictor variables**<br>
- agescrn    :  age at screening
- agecon     :  age at conception  (age at screening is used instead of age at conception)
- race       :  Race of Mother
- birthord   :  live birth order

**Depdendent  variable** or the value to be predicted (date of birth) can be derived from pregnency length
preglnth

In [7]:
import statsmodels.formula.api as smf

# build the regression formula to input the independent predictor and response /  dependent varaibles 
reg_formula = 'prglngth ~ agescrn + race + birthord'
# Create the model
reg_model = smf.ols(reg_formula, data=preg_df)
# Fit the data and find out the results
results = reg_model.fit()
# Print the sumary to view the co-efficients,  T-value, P-Value,  Std-Error etc
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:               prglngth   R-squared:                       0.003
Model:                            OLS   Adj. R-squared:                  0.003
Method:                 Least Squares   F-statistic:                     9.434
Date:                Sun, 11 Feb 2024   Prob (F-statistic):           3.20e-06
Time:                        11:53:52   Log-Likelihood:                -22060.
No. Observations:                9148   AIC:                         4.413e+04
Df Residuals:                    9144   BIC:                         4.416e+04
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     38.1241      0.177    214.862      0.000      37.776      38.472
agescrn        0.0035      0.004      0.803      0.422      -0.005       0.012
race           0.2332      0.050      4.693      0.000       0.136       0.331
birthord      -0.0575      0.027     -2.099      0.036      -0.111      -0.004
==============================================================================
Omnibus:                     6052.368   Durbin-Watson:                   1.657
Prob(Omnibus):                  0.000   Jarque-Bera (JB):           132754.618
Skew:                          -2.847   Prob(JB):                         0.00
Kurtosis:                      20.773   Cond. No.                         219.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

- agescrn    :  age at screening  attribute's co-eff,  t-value and P-value : indicate it is NOT more statistically significant.
            co-eff is very very less,  t- value is less and P-value is above 0.05,  only standard error indicate its significance
- race       :  Race of Mother attribute's co-eff,  t value and P-value : indicate it is  more statistically significant as 
            co-eff  has a 20% infleunce (0.22) ,  t-value of 4+ indicates strong assoiation_ and P value is less than 0.05,  only standard error is more than other predictors
- birthord   :  live birth order co-ef, Std-err t-value - indicate it has some statistically significant.
            co-eff is very very less i.e. negative,  t-value is again very  less but P-value is less than 0.05 showing some significance,  also standard error indicate its significance
            
------

**Exercise 11-3**

If the quantity you want to predict is a count, you can use Poisson regression, which is implemented in StatsModels with a function called poisson. It works the same way as ols and logit. As an exercise, let’s use it to predict how many children a woman has born; in the NSFG dataset, this variable is called numbabes.

Suppose you meet a woman who is 35 years old, black, and a college graduate whose annual household income exceeds $75,000. How many children would you predict she has born?

From the Excercise question following are identified<br>
**response variable** <br>
    number of childern  -  use attribute  'numbabes' as response variable

**Independent variables or Predictor**<br>
    Age          : 35 years -   use attribute 'ager' - age of respondent<br>
    Education    : college graduate - use attribute 'educat' - Education of respondent<br>
    race         : black - use attribute 'race' <br>
    annual income: $75000  - use 'totincr'  - Totsal income of respondent's family<br>

In [8]:
# Load the Respondent dta
resp_df = nsfg.ReadFemResp()

In [9]:
resp_df.shape

(7643, 3087)

In [10]:
preg_df.shape

(13593, 244)

In [11]:
# Since the numbabes,  totincr are in respondent dataframe, join both dataframe using caseid as the join condition
Preglive_resp_df = live_df.join(resp_df,  on='caseid', rsuffix='_c')

In [12]:
Preglive_resp_df.shape

(9148, 3331)

In [13]:
# Check the numbabes feature values 
Preglive_resp_df.numbabes

0        1.0
1        1.0
2        1.0
3        1.0
4        1.0
        ... 
13581    NaN
13584    NaN
13588    NaN
13591    NaN
13592    NaN
Name: numbabes, Length: 9148, dtype: float64

In [14]:
Preglive_resp_df.numbabes.isnull().sum()

3497

In [16]:
# fill numbabes with median value
import numpy as np
numbabe_med = Preglive_resp_df.numbabes.median()
Preglive_resp_df['numbabes'] = Preglive_resp_df['numbabes'].replace(np.nan, numbabe_med)

In [17]:
#Preglive_resp_df = Preglive_resp_df.dropna(subset=['age_r','educat','race','totincr'])

In [18]:
Preglive_resp_df.shape

(9148, 3331)

In [19]:
# Build the regression formula with the independent variables and dependent variable
pos_formula = 'numbabes ~ age_r + C(race) + totincr + educat'
# Create the posson distribution model as prediction is needed to find number of babies from the existing data
pos_model = smf.poisson(pos_formula, data=Preglive_resp_df)
#  Fit the variable data and retrieve the results
pos_results = pos_model.fit()
# Print the summary to find which variable has statisticaly significant impact on dependent variable
pos_results.summary()

Optimization terminated successfully.
         Current function value: 1.393384
         Iterations 6


<class 'statsmodels.iolib.summary.Summary'>
"""
                          Poisson Regression Results                          
==============================================================================
Dep. Variable:               numbabes   No. Observations:                 5651
Model:                        Poisson   Df Residuals:                     5645
Method:                           MLE   Df Model:                            5
Date:                Sun, 11 Feb 2024   Pseudo R-squ.:                  0.1073
Time:                        11:54:36   Log-Likelihood:                -7874.0
converged:                       True   LL-Null:                       -8820.6
Covariance Type:            nonrobust   LLR p-value:                     0.000
================================================================================
                   coef    std err          z      P>|z|      [0.025      0.975]
--------------------------------------------------------------------------------
Intercept       -1.2055      0.082    -14.686      0.000      -1.366      -1.045
C(race)[T.2]    -0.0256      0.028     -0.921      0.357      -0.080       0.029
C(race)[T.3]    -0.1977      0.050     -3.927      0.000      -0.296      -0.099
age_r            0.0616      0.002     40.209      0.000       0.059       0.065
totincr         -0.0520      0.003    -16.955      0.000      -0.058      -0.046
educat          -0.0067      0.005     -1.406      0.160      -0.016       0.003
================================================================================
"""

In [21]:
import pandas as pd
# predicting number of babies for woman who is 35 years old, black, and a college graduate whose annual household income exceeds $75,000
pred_columns = ['age_r', 'race', 'totincr', 'educat']
# total income - totincr  value is 14 for $75000 and above
# educat -  4 years college education value is 16
# race -  Black has value 1
pred_df = pd.DataFrame([[35, 1, 14, 16]], columns=pred_columns)
pos_results.predict(pred_df)

0    1.124008
dtype: float64

**Number of babies predicted with given dependent variables in 1.12 babies , can be rounded to 1 baby**

In [22]:
#  Add square of age of respondent to have a better prediction for non-linear relationship with numbabes
Preglive_resp_df['age_sq'] = Preglive_resp_df.age_r**2

In [23]:
# Build the regression formula with the independent variables including a derived age square variable and dependent variable
pos_formula_new = 'numbabes ~ age_r + age_sq + C(race) + totincr + educat'
pos_model_new = smf.poisson(pos_formula_new, data=Preglive_resp_df)
pos_results_new = pos_model_new.fit()
pos_results_new.summary()

Optimization terminated successfully.
         Current function value: 1.318696
         Iterations 7


<class 'statsmodels.iolib.summary.Summary'>
"""
                          Poisson Regression Results                          
==============================================================================
Dep. Variable:               numbabes   No. Observations:                 5651
Model:                        Poisson   Df Residuals:                     5644
Method:                           MLE   Df Model:                            6
Date:                Sun, 11 Feb 2024   Pseudo R-squ.:                  0.1552
Time:                        11:55:02   Log-Likelihood:                -7451.9
converged:                       True   LL-Null:                       -8820.6
Covariance Type:            nonrobust   LLR p-value:                     0.000
================================================================================
                   coef    std err          z      P>|z|      [0.025      0.975]
--------------------------------------------------------------------------------
Intercept       -7.5031      0.260    -28.879      0.000      -8.012      -6.994
C(race)[T.2]    -0.0042      0.028     -0.153      0.878      -0.059       0.050
C(race)[T.3]    -0.1344      0.050     -2.668      0.008      -0.233      -0.036
age_r            0.4799      0.016     30.062      0.000       0.449       0.511
age_sq          -0.0066      0.000    -26.715      0.000      -0.007      -0.006
totincr         -0.0542      0.003    -17.568      0.000      -0.060      -0.048
educat          -0.0066      0.005     -1.383      0.167      -0.016       0.003
================================================================================
"""

In [24]:
# predicting number of babies for woman who is 35 years old, black, and a college graduate whose annual household income 
# exceeds $75,000 and including square of ages derived value for better non-linear prediction
pred_columns = ['age_r', 'age_sq', 'race', 'totincr', 'educat']
# total income - totincr  value is 14 for $75000 and above
# educat -  4 years college education value is 16
# race -  Black has value 1
# age is 35
# age_sq value is square of 35
pred_df = pd.DataFrame([[35, 35**2, 1, 14, 16]], columns=pred_columns)
pos_results_new.predict(pred_df)

0    1.468206
dtype: float64

**Number of babies predicted with given dependent variables in 1.47 babies or 1.5 babies (Woman can have 1 or 2 babies)**

----

**Exercise 11-4**

If the quantity you want to predict is categorical, you can use multinomial logistic regression, which is implemented in StatsModels with a function called mnlogit. As an exercise, let’s use it to guess whether a woman is married, cohabitating, widowed, divorced, separated, or never married; in the NSFG dataset, marital status is encoded in a variable called rmarital.

Suppose you meet a woman who is 25 years old, white, and a high school graduate whose annual household income is about $45,000. What is the probability that she is married, cohabitating, etc?

From the Excercise question following are identified<br>
**response variable** is martial status -  attribute is 'rmartial'

**Independent variables**<br>
    Age          : 25 years -   use attribute 'age_r' - age of respondent<br>
    Education    : high school graduate - use attribute 'educat' - Education of respondent<br>
    race         : white - use attribute 'race' <br>
    annual income: $45000  - use 'totincr'  - Totsal income of respondent's family<br>


In [25]:
# Build the regression formula with the independent variables and dependent variable
logit_formula='rmarital ~ age_r  + C(race) + totincr + educat'
# Create a multinomial logit regrssion  model as prediction is needed to find marital status from the existing data
# Since a categorical non-binary classification is required to be predicted multinomial logit regrssion is used 
logit_model = smf.mnlogit(logit_formula, data=Preglive_resp_df)
#  Fit the variable data and retrieve the results
logit_results = logit_model.fit()
# Print summary to find the significance impact of independent variables for each categorical classification response variable values
logit_results.summary() 

Optimization terminated successfully.
         Current function value: 1.188934
         Iterations 8


<class 'statsmodels.iolib.summary.Summary'>
"""
                          MNLogit Regression Results                          
==============================================================================
Dep. Variable:               rmarital   No. Observations:                 5651
Model:                        MNLogit   Df Residuals:                     5621
Method:                           MLE   Df Model:                           25
Date:                Sun, 11 Feb 2024   Pseudo R-squ.:                 0.08755
Time:                        11:55:21   Log-Likelihood:                -6718.7
converged:                       True   LL-Null:                       -7363.3
Covariance Type:            nonrobust   LLR p-value:                1.647e-256
================================================================================
  rmarital=2       coef    std err          z      P>|z|      [0.025      0.975]
--------------------------------------------------------------------------------
Intercept        2.5003      0.324      7.709      0.000       1.865       3.136
C(race)[T.2]    -1.1476      0.106    -10.829      0.000      -1.355      -0.940
C(race)[T.3]    -0.7534      0.170     -4.426      0.000      -1.087      -0.420
age_r           -0.0075      0.005     -1.384      0.166      -0.018       0.003
totincr          0.0347      0.012      2.889      0.004       0.011       0.058
educat          -0.2789      0.021    -13.298      0.000      -0.320      -0.238
--------------------------------------------------------------------------------
  rmarital=3       coef    std err          z      P>|z|      [0.025      0.975]
--------------------------------------------------------------------------------
Intercept       -0.9400      0.955     -0.984      0.325      -2.813       0.933
C(race)[T.2]    -0.8189      0.313     -2.619      0.009      -1.432      -0.206
C(race)[T.3]    -1.1772      0.638     -1.844      0.065      -2.428       0.074
age_r            0.0113      0.016      0.689      0.491      -0.021       0.044
totincr         -0.0090      0.035     -0.255      0.799      -0.078       0.060
educat          -0.2202      0.061     -3.615      0.000      -0.340      -0.101
--------------------------------------------------------------------------------
  rmarital=4       coef    std err          z      P>|z|      [0.025      0.975]
--------------------------------------------------------------------------------
Intercept       -0.8573      0.310     -2.764      0.006      -1.465      -0.249
C(race)[T.2]    -0.5505      0.115     -4.771      0.000      -0.777      -0.324
C(race)[T.3]    -0.7380      0.203     -3.644      0.000      -1.135      -0.341
age_r            0.0005      0.005      0.092      0.927      -0.010       0.011
totincr          0.0085      0.012      0.715      0.475      -0.015       0.032
educat          -0.0396      0.018     -2.177      0.029      -0.075      -0.004
--------------------------------------------------------------------------------
  rmarital=5       coef    std err          z      P>|z|      [0.025      0.975]
--------------------------------------------------------------------------------
Intercept        0.0815      0.389      0.210      0.834      -0.680       0.843
C(race)[T.2]    -1.3716      0.127    -10.817      0.000      -1.620      -1.123
C(race)[T.3]    -0.4130      0.182     -2.274      0.023      -0.769      -0.057
age_r           -0.0040      0.007     -0.602      0.547      -0.017       0.009
totincr          0.0926      0.015      5.980      0.000       0.062       0.123
educat          -0.1634      0.024     -6.802      0.000      -0.210      -0.116
--------------------------------------------------------------------------------
  rmarital=6       coef    std err          z      P>|z|      [0.025      0.975]
--------------------------------------------------------------------------------
Intercept        4.2030      0.301     13.967      0.000       3.613       4.793
C

In [26]:
# Build a input datframe with the given inputs 
pred_logit_columns = ['age_r',  'race', 'totincr', 'educat']
# total income - totincr  value is 11 for $45000 income
# educat -  high school diploma education value is 12
# race -  White has value 1
# age is 25
pred_logit_df = pd.DataFrame([[25, 2, 11, 12]], columns=pred_logit_columns)
logit_results.predict(pred_logit_df)

0         1         2      3         4         5
0  0.632409  0.104414  0.009325  0.107  0.061406  0.085446

**The prediction is**<br>
The woman's marital status is  <br>
- 'Currently Married' with 63% probability<br>
- 'Not Married but living with opp sex partner' with 10% probability<br>
- 'Divorced' with 11% probability<br>
- 'Never been Married' with 8% probability<br>

In [27]:
# Build the regression formula with the independent variables, including squared values of age to aid non-linear relationship
# predictions and dependent variable
logit_formula='rmarital ~ age_r + age_sq + C(race) + totincr + educat'
# Create a multinomial logit regrssion  model
logit_model = smf.mnlogit(logit_formula, data=Preglive_resp_df)
#  Fit the variable data and retrieve the results
logit_results = logit_model.fit()
# Print summary to find the significance impact of independent variables for each categorical classification response variable values
logit_results.summary() 

Optimization terminated successfully.
         Current function value: 1.187797
         Iterations 8


<class 'statsmodels.iolib.summary.Summary'>
"""
                          MNLogit Regression Results                          
==============================================================================
Dep. Variable:               rmarital   No. Observations:                 5651
Model:                        MNLogit   Df Residuals:                     5616
Method:                           MLE   Df Model:                           30
Date:                Sun, 11 Feb 2024   Pseudo R-squ.:                 0.08842
Time:                        11:55:26   Log-Likelihood:                -6712.2
converged:                       True   LL-Null:                       -7363.3
Covariance Type:            nonrobust   LLR p-value:                5.114e-255
================================================================================
  rmarital=2       coef    std err          z      P>|z|      [0.025      0.975]
--------------------------------------------------------------------------------
Intercept        2.4450      0.652      3.748      0.000       1.166       3.724
C(race)[T.2]    -1.1473      0.106    -10.826      0.000      -1.355      -0.940
C(race)[T.3]    -0.7538      0.170     -4.428      0.000      -1.088      -0.420
age_r           -0.0035      0.041     -0.085      0.932      -0.085       0.078
age_sq       -6.911e-05      0.001     -0.099      0.921      -0.001       0.001
totincr          0.0347      0.012      2.892      0.004       0.011       0.058
educat          -0.2788      0.021    -13.291      0.000      -0.320      -0.238
--------------------------------------------------------------------------------
  rmarital=3       coef    std err          z      P>|z|      [0.025      0.975]
--------------------------------------------------------------------------------
Intercept        0.4416      1.925      0.229      0.819      -3.330       4.214
C(race)[T.2]    -0.8215      0.313     -2.626      0.009      -1.435      -0.208
C(race)[T.3]    -1.1859      0.638     -1.857      0.063      -2.437       0.066
age_r           -0.0882      0.122     -0.721      0.471      -0.328       0.152
age_sq           0.0017      0.002      0.819      0.413      -0.002       0.006
totincr         -0.0095      0.035     -0.270      0.787      -0.078       0.059
educat          -0.2208      0.061     -3.620      0.000      -0.340      -0.101
--------------------------------------------------------------------------------
  rmarital=4       coef    std err          z      P>|z|      [0.025      0.975]
--------------------------------------------------------------------------------
Intercept       -1.1232      0.655     -1.714      0.087      -2.408       0.161
C(race)[T.2]    -0.5498      0.115     -4.764      0.000      -0.776      -0.324
C(race)[T.3]    -0.7357      0.203     -3.632      0.000      -1.133      -0.339
age_r            0.0197      0.042      0.469      0.639      -0.063       0.102
age_sq          -0.0003      0.001     -0.461      0.645      -0.002       0.001
totincr          0.0087      0.012      0.734      0.463      -0.015       0.032
educat          -0.0395      0.018     -2.170      0.030      -0.075      -0.004
--------------------------------------------------------------------------------
  rmarital=5       coef    std err          z      P>|z|      [0.025      0.975]
--------------------------------------------------------------------------------
Intercept        2.0479      0.767      2.670      0.008       0.545       3.551
C(race)[T.2]    -1.3781      0.127    -10.853      0.000      -1.627      -1.129
C(race)[T.3]    -0.4229      0.182     -2.326      0.020      -0.779      -0.067
age_r           -0.1470      0.049     -3.010      0.003      -0.243      -0.051
age_sq           0.0024      0.001      2.953      0.003       0.001       0.004
totincr          0.0914      0.015      5.907      0.000       0.061       0.122
educat          -0.1654      0.024     -6.864      0.000      -0.213      -0.118
-

In [28]:
# predicting marital status of a woman who is 25 years old, white, and a school diploma graduate whose annual household income is $45,000
pred_logit_columns = ['age_r', 'age_sq', 'race', 'totincr', 'educat']
# total income - totincr  value is 11 for $45000 income
# educat -  high school diploma education value is 12
# race -  White has value 1
# age is 25
# age_sq value is square of 25
pred_logit_df = pd.DataFrame([[25, 25**2, 2, 11, 12]], columns=pred_logit_columns)
logit_results.predict(pred_logit_df)

0         1         2         3         4         5
0  0.634291  0.105153  0.008592  0.109129  0.053474  0.089361

**No change noticed in predicted probabilites ,  even if the age value is squared and added as another independent variable for the response variable**

**The prediction is**<br>
The woman's marital status is  <br>
- 'Currently Married' with 63% probability<br>
- 'Not Married but living with opp sex partner' with 10% probability<br>
- 'Divorced' with 11% probability<br>
- 'Never been Married' with 9% probability<br>